docs/tutorial/writing-tests/test-case.rst

```diff
@@ -0,0 +1,1148 @@
+  test for the loopback interface)
+- The resources that, when filtered by the ``resource`` expression is empty,
+  marks the job as skipped. We do this when the job makes sense (for example,
+  the interface exists) but the current situation makes it impossible for it
+  to pass for an external reason (for example, the ethernet port may work but
+  it is not currently plugged in)
+
+Dealing with complexity - Python
+================================
+
+The ``network_available`` test that we have created during this tutorial is
+very simple but, in the real world things are not as simple. For example,
+right now we are only pinging once from the test, if the ping goes through
+we call that a success, else a failure. This works in our simple scenario while
+developing the test, but when hundreds of devices all try to ping at the same
+time things can get messy quickly, and messages can get lost. One possible
+evolution for this test is to do more pings and use the packet
+loss output to decide if we can call the test a success or a failure.
+
+Translating the test to Python
+------------------------------
+
+While we could do this with a tall jenga tower entirely constituted of pipes,
+tee and ``awk`` commands, always keep in mind, the best foot gun is the one we
+don't use. Checkbox allows you to write hundreds of lines of code in the
+command section but this doesn't make it a good idea. When we need to evolve
+beyond a few lines of bash we always suggest a rewrite in Python and to add
+proper unit tests.
+
+.. note::
+    While there is no formal rule on the maximum size or complexity of a
+    command section, as a rule of thumb avoid using nested ifs/for loops,
+    multiple pipes and destructive redirection within a command section. You
+    will thank us later.
+
+Create two new directories in the provider: ``bin/`` and ``tests/``. Create
+a new python file in ``bin/`` and call it ``network_available.py`` and make it
+executable (``chmod +x network_available.py``).
+
+Let's translate the previous test into Python first:
+
+.. code-block:: python
+
+    #!/usr/bin/env python3
+    import sys
+    import argparse
+    import subprocess
+
+
+    def parse_args(argv):
+        parser = argparse.ArgumentParser()
+        parser.add_argument(
+            "interface", help="Interface to connectivity test"
+        )
+        return parser.parse_args(argv)
+
+
+    def network_available(interface):
+        print("Testing", interface)
+        return subprocess.check_call(
+            ["ping", "-I", interface, "-c", "1", "1.1.1.1"]
+        )
+
+
+    def main(argv=None):
+        if argv is None:
+            argv = sys.argv[1:]
+        args = parse_args(argv)
+        ping_test(args.interface)
+
+
+    if __name__ == "__main__":
+        main()
+
+.. note::
+    A few important things to notice about the script:
+
+    #. We use Black to format all tests and source files in Checkbox with a custom config: ``line-length = 79``.
```

Comment:
```suggestion
    - We use Black to format all tests and source files in Checkbox with a custom config: ``line-length = 79``.
```
Ordering is not important in the list, so bullets would be better (which is used in a note below)